### PR TITLE
Hw3 testing

### PR DESCRIPTION
**Pull request by Ndm1t**
Added unit tests for controllers services and other infrastructure layer structures + added integration tests for /api/rate and /api/subscribe

### CODE REVIEW

**Review by irynakopyl — COMMENTED** [review]
image
you can change base branch to the hw2, so as a result there will be only changed files in this brach

File: internal/server/controllers/reateController_test.go
Comment (irynakopyl) [inline internal/server/controllers/reateController_test.go:1]: typo in the name of the file r _e_ ateController

**Review by irynakopyl — COMMENTED** [review]
General reccomendation is to move tests to the separate directory

File: internal/database/subscriptionDao_test.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	})
+	t.Run("update subscription success", func(t *testing.T) {
+		email := "test@example.com"
+		foundSubscription, err := dao.Find(email)
+		require.NoError(t, err)
+
+		foundSubscription.Status = models.Unsubscribed
+		updatedSubscription, err := dao.Update(*foundSubscription)
+		require.NoError(t, err)
+		assert.Equal(t, models.Unsubscribed, updatedSubscription.Status)
```
Comment (RuslanMelnychenko) [inline internal/database/subscriptionDao_test.go:62]: You can also additionally make a check for getting the `.Find()` of the changed record. To be sure that subsequent `.Find()` queries will return the changed record.

**Review by RuslanMelnychenko — APPROVED** [review]

**Comment by Ndm1t** [comment]
> General reccomendation is to move tests to the separate directory

In go it is common practice to have test files next to the file with the tested functionality, was quite unusual to me after JS as well)

**Pull request merged**